src/storage/src/render/upsert.rs

```diff
@@ -213,18 +255,31 @@ where
                     // along with the value with the _latest timestamp for that key_.
                     commands_state.clear();
                     for (_, key, _, _) in stash.iter().take(idx) {
-                        commands_state
-                            .entry(*key)
-                            .or_insert_with(|| state.get(key).cloned());
+                        commands_state.entry(*key).or_insert(None);
                     }
 
+                    // These iterators iterate in the same order because `commands_state`
+                    // is an `IndexMap`.
+                    multi_get_scratch.clear();
+                    multi_get_scratch.extend(commands_state.iter().map(|(k, _)| *k));
+                    state
+                        .multi_get(
+                            multi_get_scratch.drain(..),
+                            commands_state.iter_mut().map(|(_, v)| v),
+                        )
+                        .await
+                        .expect("hashmap impl to not fail");
+
                     // From the prefix that can be emitted we can deduplicate based on (ts, key) in
                     // order to only process the command with the maximum order within the (ts,
                     // key) group. This is achieved by wrapping order in `Reverse(order)` above.
-                    let mut commands = stash.drain(..idx).dedup_by(|a, b| {
-                        let ((a_ts, a_key, _, _), (b_ts, b_key, _, _)) = (a, b);
-                        a_ts == b_ts && a_key == b_key
-                    });
+                    let mut commands = stash
+                        .drain(..idx)
+                        .dedup_by(|a, b| {
+                            let ((a_ts, a_key, _, _), (b_ts, b_key, _, _)) = (a, b);
+                            a_ts == b_ts && a_key == b_key
+                        })
+                        .peekable();
```

Comment:
Why `peekable()`? (I did not see any `peek` calls, is there any behaviour change by using this?)

Reply:
Ditto, it's probably a leftover from the previous iteration of this PR